src/main/java/net/snowflake/client/core/bind/BindUploader.java

```diff
@@ -189,13 +196,126 @@ public static synchronized BindUploader newInstance(SFSession session, String st
    * @param bindValues the bind map to upload
    * @throws BindException if the bind map could not be serialized or upload fails
    */
-  public void upload(Map<String, ParameterBindingDTO> bindValues) throws BindException {
+  @Deprecated
+  public void uploadDeprecated(Map<String, ParameterBindingDTO> bindValues) throws BindException {
     if (!closed) {
       serializeBinds(bindValues);
       putBinds();
     }
   }
 
+  /**
+   * Upload bind parameters via streaming. This replaces previous function upload function where
+   * binds were written to a file which was then uploaded with a PUT statement.
+   *
+   * @param bindValues the bind map to upload
+   * @throws BindException if the bind map could not be serialized or upload fails
+   * @throws SQLException if the upload fails
+   */
+  public void upload(Map<String, ParameterBindingDTO> bindValues)
+      throws BindException, SQLException {
+    if (!closed) {
+      List<ColumnTypeDataPair> columns = getColumnValues(bindValues);
+      List<byte[]> csvRows = buildRowsAsBytes(columns);
+      int startIndex = 0;
+      int numBytes = 0;
+      int rowNum = 0;
+      int fileCount = 0;
+
+      while (rowNum < csvRows.size()) {
+        // create a list of byte arrays
+        while (numBytes < inputStreamBufferSize && rowNum < csvRows.size()) {
+          numBytes += csvRows.get(rowNum).length;
+          rowNum++;
+        }
+        // concatenate all byte arrays into 1 and put into input stream
+        ByteBuffer bb = ByteBuffer.allocate(numBytes);
+        for (int i = startIndex; i < rowNum; i++) {
+          bb.put(csvRows.get(i));
+        }
+        byte[] finalBytearray = bb.array();
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(finalBytearray);
+        // do the upload
+        String fileName = Integer.toString(++fileCount);
+        uploadStreamInternal(inputStream, fileName, true);
+        startIndex = rowNum;
+        numBytes = 0;
+        try {
+          inputStream.close();
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+      }
+    }
+  }
+
+  /**
+   * Method to put data from a stream at a stage location. The data will be uploaded as one file. No
+   * splitting is done in this method. Similar to uploadStreamInternal() in SnowflakeConnectionV1.
+   *
+   * <p>Stream size must match the total size of data in the input stream unless compressData
+   * parameter is set to true.
+   *
+   * <p>caller is responsible for passing the correct size for the data in the stream and releasing
+   * the inputStream after the method is called.
+   *
+   * @param inputStream input stream from which the data will be uploaded
+   * @param destFileName destination file name to use
+   * @param compressData whether compression is requested fore uploading data
+   * @throws SQLException raises if any error occurs
+   */
+  private void uploadStreamInternal(
```

Comment:
Very similar to function in SnowflakeConnectionV1. Uses a put statement and inputstream to upload.

Reply:
Sorry, I have a small question about the whole work flow. So to my understanding, when inserting a huge amount of data into a table, the data can be uploaded by stream as multiple files in our temporary stage. Then how can those data (in multiple uploaded files) get into the table finally?

Reply:
After the data is uploaded as multiple files on the internal stage, it is moved into tables with the COPY INTO command. This part of the flow doesn't have to change at all. The altered part is the uploading to the internal stage.